### ISSUE TITLE
TVTB: The VCF Tool Box

### ISSUE THREAD

**Issue by kevinrue** [Bioconductor/Contributions#111]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/kevinrue/TVTB

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by kevinrue** [comment]
Dear all,

I'm not sure whether there is a way to avoid the *nowidow.sty* LATEX package or if *moscato1* needs a little update.

Note that I had a look at the *tex document and I cannot see any occurence of "nowidow" in it. 

Kind regards,
Kevin

**Comment by mtmorgan** [comment]
Hi Kevin -- probably it is an indirect LaTeX dependency; we updated the build server to have the package. I'll review it as-is.

**Comment by lawremi** [comment]
This looks like an important contribution, and I'd like to provide constructive feedback. Would it be best to post them as issues on the source repository, or here? I guess I will post issues and reference them from here, if that's even possible. What's the best practice here?

**Comment by mtmorgan** [comment]
It is helpful to have your comments included in this issue, so at least a summary or link to issues / pull requests made elsewhere is appropriate.

**Comment by kevinrue** [comment]
Hi Michael,

I agree with Martin. Moreover, this is also what I expected from reading the submission [guidelines](https://github.com/Bioconductor/Contributions#what-to-expect)<sup>1</sup>, and that I hoped for by submitting the package now, before moving on to even more ambitious features. 

> <sup>1</sup>Other Bioconductor developers and users with domain expertise are encouraged to provide additional community commentary. Reviewers will add comments to the issue you created.

Even if it means "missing" the next release (interested users could always get it from the *devel* branch), I am happy to take onboard all the constructive feedback that I can get to validate the new concepts I wish to contribute, or rectify my trajectory where necessary; I have particularly in mind the *VCF filter rules* and to a lesser extent the more experimental `TVTBparam` class.

Thanks for your interest!

**Comment by lawremi** [comment]
I find these linear threads to be tough to follow when discussion becomes complex.

But, just to get thing started, I find the explicit argument naming in the method and class definitions to impinge on readability. Perhaps a matter of personal taste/habit. But it's tough for me to read.

You might not need to define so many `initialize()` methods. Combined with the class prototype, the default method usually works. It's such a low-level generic, we usually leave it alone and instead define high-level constructor functions.

It also seems you could make better use of dispatch to handle the polymorphic behaviors of the different filter types. E.g., you could avoid stuff like `switch()` statements. It's a more extensible approach that might also clarify the code.

**Comment by kevinrue** [comment]
I see all your points.

Regarding the linear thread, I think it's good to mention here for starters. We can always open issues for fixes or enhancements that require a more in-depth discussion.

I can see all your points, but I'll need a little more details to focus my updates:
1. argument naming: could you please list the ones that you have in mind? 
2. `initialize` methods: I'll look back into that using some other BioC packages as template. To be honest, this is the first time I implemented S4 classes following online tutorials and I might have gone over the top following them blindly in some places. I can't find right now where I got the `initialize` concept from, but I'm happy to remove it if that's not standard practice.
3. I see the `switch` statements that you mean, and I think I can see how to replace them by a method dispatch.

So please just clarify point 1, and keep the feedback coming; I'm happy to feel some interest :)
Best wishes,
Kevin

**Comment by lawremi** [comment]
I just mean that naming every argument to `setMethod()` makes it hard to parse. That might just be me being used to the standard syntax for method definition.

**Comment by kevinrue** [comment]
So we're talking about changing:

    setMethod(
        f = "addCountGenos",
        signature = c(vcf="ExpandedVCF"),
        definition = function(
            vcf, genos, key, description, samples = 1:ncol(vcf), force = FALSE){
            <code>
            }
into:

    setMethod(
        "addCountGenos", c("ExpandedVCF"),
        function(
            vcf, genos, key, description, samples = 1:ncol(vcf), force = FALSE){
            <code>
            }

Right? (Just making sure before I update the whole package on a misunderstanding :grin: )

Spelling out the definition of S4 methods and classes was a bit clearer to me while learning the meaning of each argument, but I think I got familiar enough now to move on to the latter syntax, which seems to be the Bioconductor standard as far as I can see (S4Vectors, VariantAnnotation, ...).

Cheers!

**Comment by lawremi** [comment]
Yea, usually it's like:

```
setMethod("addCountGenos", "ExpandedVCF", function(...) {
})
```

**Comment by kevinrue** [comment]
HI Michael,

I've created the branch [BiocReview](https://github.com/kevinrue/TVTB/tree/BiocReview) to work on your comments.
So far, I've dealt with the `setMethod()` point, if you prefer to checkout the code from that branch for reviewing purpose.

Going through the code for the above point, I have identified some code cleaning that will make it easier to read and more organised before dealing with your other two points (I call them `initialize` and `switch`)

All the best,
Kevin

**Comment by kevinrue** [comment]
Hi @lawremi ,

> It also seems you could make better use of dispatch to handle the polymorphic behaviors of the different filter types. E.g., you could avoid stuff like switch() statements. It's a more extensible approach that might also clarify the code.

I have replaced two of the four `switch` statements in *VcfFilterRules-class.R*, by implementing accessors that return `NA_character_` for filter classes that do not have the relevant slot (*e.g.* `VcfFixedRules` do not have `vep` and `type` slot; the accessors return NA_character_ to provide a consistent type of return value).

Looking back at the `switch` statement in `[<-` for signature `VcfFilterRules`, I realise that I wrote those complex methods to override the behaviour of the parent method (inherited from `List`), which does not transfer the `name` and `active` data of rules (it does transfer the `elementMetadata` though). I sent an email to maintainer at bioconductor to discuss this outside this linear thread.

I think my mechanism could be greatly simplified by the following
* I should probably move data from the `type` slot of `VcfFilterRules` to `elementMetadata` to solve part of the problem:
  + `elementMetadata` must be the same length as the number of rules; which matches the fact that there must be as many "type" values as rules.
  + `[` and `[<-` handle `elementMetadata`, which I wouldn't have to deal with anymore
* I wonder whether the `vep` slot of {`VcfVepRules` and `VcfFilterRules`} would be better stored in the `metadata` slot , rather than creating a new slot for this `character` vector of length 1 ?
  + I could just have my class validity check that `identical(length(metadata(x)[["vep]]), 1)`

In any case, I need to re-design my filter classes, and their subsetting/replacement methods, all depending on the parent `FilterRules` class and associated methods.

Thanks in advance for advising,
Kevin

**Comment by lawremi** [comment]
I wouldn't use any of the metadata components to support a formal representation. You can set a method on the `parallelSlotNames()` generic to handle your `type` slot.

**Comment by kevinrue** [comment]
Hi @lawremi ,

Just to let you know that following your advice, I considerably trimmed down my code. 

* I have removed definitons of `initialize` methods, and moved the code to high level constructor methods (latest relevant commit:  [89bb2b3](https://github.com/kevinrue/TVTB/commit/94a48e8522df5bad85e44022e0354bb16adf7a67))
* I greatly simplified the `[` and [<-` override methods for VCF filter rules, applying the parent behaviour for all inherited slots (latest relevant commit: [0da247b](https://github.com/kevinrue/TVTB/commit/92b17a617c6bac0135bbed2a691f72b792321a4b)):
  + for consistency, to avoid confusing users
  + for propagation of fixes from the parent classes to the child class

Thanks for the constructive feedback! It is great to see my code being "simplified" by those changes.
Also. revisiting my code, I noticed some issues, including `if` blocks that span multiple lines without being enclosed in `{}`. I'm fixing that while awaiting Martin's review (I assume based on the version of the code that I originally submitted).

**Comment by lawremi** [comment]
Thanks for these efforts. It's exciting that we will soon gain a set of high quality tools sitting downstream of VEP-based workflows.

Here are some more suggestions:
1. Would it make sense to more formally define the `genos` slot of _TVTBparam_?
2. In the introduction vignette, when importing the phenotypes, you can pass multiple path tokens to `system.file()`, so there is no need for a separate call to `file.path()`.
3. Would it make sense to define a coercion, via `setAs()`, to generate a _ScanVcfParam_ from a _TVTBparam_ ? And maybe add a `readVcf()` method on _TVTBparam_ for convenience? That method might also accept the phenotype information.
4. I really like how you have factored all of the options into the _TVTBparam_ class. What attaching the _TVTBparam_ to the VCF object returned from your (potential) `readVcf()` wrapper? You could use the `metadata()`. Then, the user would not need to pass the `param` to downstream functions.
5. There are some summary functions that have a `plot` argument for generating a plot. It would be cleaner if the summary was separated from the plot, so I could generate a summary and later plot it. Perhaps a high-level object could be returned that has a method on the ggplot2 `autoplot()` generic.

**Comment by kevinrue** [comment]
Hi Michael,

Thanks for the encouragement. I really appreciate a community feedback to distinguish generally useful features from those specific to my specific interests.

I am still going through the methods to get rid of superfluous explicit argument naming. But I'd appreciate if you could clarify (maybe edit your post instead of posting another one?):

**(1)** Do you mean an S4 class dedicated for this purpose? Now that I think of it, it would probably be clearer/cleaner to have a dedicated validity method, rather than including this in the `TVTBparam` validity method.

**(3)** Interesting thought. I [used to have](https://github.com/kevinrue/TVTB/commit/9d8197df4c1f02bdd177fa3702ad7ad08affc413#diff-1246b373486cca4cc36b381df6d46241) a `preprocessVcf` method wrapped around `readVcf` which was handling VCF file and phenotype, and ended up with exponential increase of the number of signatures (`phenotypes={data.frame, DataFrame,missing}`, `vcf={character,TabixFile}`). I gave up the idea thinking that users may prefer to handle the few lines of code to import and combine data. It might be worth revisiting the idea after I finish my current cleaning.

**(4)** It did cross my mind. I think I was just shy about touching `VCF` objects. Worth revisiting after point **(3)** above.

**(5)** Agreed. I have recently banged my head against the wall figuring out ways to make the `*ByPhenotype` methods cleanly use loops around their associated `*InPhenotype` counterpart. Moving the `plot` argument out of the equation would definitely simplify things by avoiding unnecessary if-else blocks. I didn't think of the `autoplot` method!

**Comment by lawremi** [comment]
Yea, I meant an S4 class for (1). For (3), you might not need so many top-level signatures. Instead, your implementation could be abstracted by dispatch through lower level generics.

**Comment by kevinrue** [comment]
If that's okay, I'll use (update) this post to reflect progress on your five suggestions, now that I'm pretty clear about them.

1. 🔄 Formally define the `genos` slot of `TVTBparam`
  + 🔄I am implementing an S4 class to store genotypes.
  + 🔄I am simultaneously replacing my superfluous `hRef` and `hAlt` accessors, by `ref` and `alt` using the generic already defined in `VariantAnnotation`.
  + Once ready, I will update `TVTBparam` to store an object of that class in its `genos` slot.
2. ✅  Done.
  + Smarter call to `system.file`
3. 💤  Pending **1.** *Somewhat affected by genotype storage structure*
  + `setAs()` `ScanVcfParam` from a `TVTBparam`
  +  `readVcf()` method on `TVTBparam` for convenience? Support optional phenotype information.
4. 💤  Pending **3.** *Requires TVTBparam store in the VCF*
  + Attaching `TVTBparam` to the `VCF` object returned from the above `readVcf()` wrapper. Could use the `metadata` slot.
  + Removes requirement for `param` argument in downstream functions.
5. 💤  Pending **4.**
  + Limit existing method to summarisation. Store summarised data in adequate class.
  + Implement separate downstream method for plotting. Add methods on the `ggplot::autoplot()` generic.

All the best,
Kevin

**Comment by mtmorgan** [comment]
Hi Kevin -- the formal deadline for adding packages to the forthcoming release of Bioconductor is [Wednesday, October 12][1]; are you able to complete your modifications soon? Thanks for your work on the package.

[1]: http://bioconductor.org/developers/release-schedule/

**Comment by kevinrue** [comment]
Dear Martin,

Thanks for your e-mail.
I am fairly sure that I won't make this deadline. Unfortunately, I had
limited time recently, as this is a personal endeavour, complementary to my
postdoctoral duties, yet not encouraged during my working hours for the
benefit of other duties.

I'm happy to make the code available in the next devel branch, it will give
me more time to polish the code and docs.

Note that in the next (big) commit the families of functions tabulate* and
density* will be removed (the code will be stored in a "sandbox" subfolder
until the feature is made available again) as I agree with Michael: I need
to separate summarization from plotting. With that in mind, I actually
started wondering whether it would actually be simpler to leave some of the
summarization to the user, rather than developing a complex multi-purpose
wrapper method.

I'll post again here when after the next commit for a clearer update.

Cheers
Kevin

**Comment by mtmorgan** [comment]
OK thanks for the update Kevin and for your work on this package.

**Comment by kevinrue** [comment]
Hi @mtmorgan ,

Actually, if that's okay with you, I have [merged](https://github.com/kevinrue/TVTB/pull/3) the BiocReview branch of my repository to the master branch built by the tracker. I have now reacted to all of @lawremi 's suggestions, and trimmed out unstable/unnecessary methods (see below). I believe the current features may actually be helpful to release already, while I work on the *devel* branch to experiment and develop more features.

A couple of considerations:
* As I pointed out in my previous message, I have now removed the families of methods `tabulate*` and `density*`. I have not decided yet whether this removal is temporary or permanent. I need time to experiment, but I have a feeling that those wrapper methods are more cumbersome and less flexible than a few lines of code. My plan is to re-introduce those features in future releases of the *Introduction.pdf* vignette through some "manual" code , and then decide whether a wrapper method is worth implementing.
* Again, it seems that the *moscato1* node is acting up. I don't understand why the vignette build would fail like it does, only on a single build node.

Kind regards,
Kevin

**Comment by mtmorgan** [comment]
This type of error seems to occur when the processor gets ahead of the file system, for instance because large files (e.g., vector representations of millions of points, rather than reasonable-sized bitmaps) are being created. I don't know if that offers any hints? We could / will 'let it go' now, but these tend to re-appear during the regular builds.

**Comment by kevinrue** [comment]
Actually, it may be unrelated, but I just realise now that the warning messages displayed in the errored build should not be there at all (my methods are designed to avoid raising those `VariantAnnotation` warnings). I'm homing in on the source of that problem right now, ensuing from recent updates as part of the review process, and I should be able to trigger another build in a couple of hours.

I do not see a link between those warnings and the build error, but maybe there is one. Otherwise, if it crops up again in future builds, I'll be happy to investigate in more detail. It never happened to me so far (Ubuntu & OSX machines tested)

I am only using a small VCF file as input (150kb *.vcf.gz file; 1.3 MB VCF object in memory during the R session; 481 variants found within a single gene coding region). Do you see any obvious issue in this?
I don't see where I would be generating millions of points.

**Comment by kevinrue** [comment]
Hm... The messages that appear in the new failed build are now exactly the ones expected. 

However, as I suspected, that issue was distinct from the source of the failed build on *moscato1*.
It's a shame I haven't yet a Windows machine set up to build R packages.

Is there any chance that using `BiocParallel` (even with `SerialParam`) causes any issue on Windows? I have come across situations where some OSes didn't like certain parallel settings. I remember seeing Windows mentioned quite a few times on forums about that.

Thanks for advising
Kevin

**Comment by mtmorgan** [comment]
@kevinrue yes, open connections have caused problems in the past, which can happen when a 'worker' is either not closed (`bpstart()` / `bpstop()` not balanced; if you do not explicitly call these then not relevant) or a worker dies but is not cleaned up (e.g., because the worker dies but is itself holding on to some other connection, like an open file). The worker dying would be revealed by `SerialParam()`. I'll try later today to debug this.

**Comment by kevinrue** [comment]
Hi @mtmorgan 
(@aoles might be interested in this too)

I have come across something strange, which might be related to my Windows build problem):

After setting up a virtual machine running Windows 10, Rtools.exe, Perl, MikTek 2.9, ...

`BiocStyle::pdf_document()` works:
```
render("Introduction.Rmd", BiocStyle::pdf_document(), intermediates_dir = "tmp", clean = FALSE)
```

`pdf_document2()` (which I have used so far) does not work:
```
render("Introduction.Rmd", BiocStyle::pdf_document2(), intermediates_dir = "tmp", clean = FALSE)
```
with message (after successfully evaluating all the R code blocks)

```
...
output file: C:/Users/Rue-AlbrechtKevinC/Documents/git/TVTB/vignettes/tmp/Introduction.knit.md

"C:/Program Files/RStudio/bin/pandoc/pandoc" +RTS -K512m -RTS "C:/Users/Rue-AlbrechtKevinC/Documents/git/TVTB/vignettes/tmp/Introduction.utf8.md" --to latex --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash --output pandoc1d3837299.tex --table-of-contents --toc-depth 2 --template "C:\Users\RUE-AL~1\AppData\Local\Temp\RtmpqWOvZ1\BiocStyle\template.tex" --number-sections --highlight-style tango --latex-engine pdflatex --include-in-header "C:\Users\RUE-AL~1\AppData\Local\Temp\RtmpqWOvZ1\1d385ec8499.tex" --bibliography TVTB.bib --filter pandoc-citeproc 
Error: Failed to compile Introduction.tex.
In addition: Warning messages:
1: running command '"C:\Users\RUE-AL~1\AppData\Local\Programs\MIKTEX~1.9\miktex\bin\x64\latexmk.exe" -pdf -latexoption=-halt-on-error -interaction=batchmode -pdflatex="pdflatex" "Introduction.tex"' had status 1 
2: running command '"C:\Users\RUE-AL~1\AppData\Local\Programs\MIKTEX~1.9\miktex\bin\x64\latexmk.exe" -v' had status 1 
```

To test my working hypothesis, I thing I will commit a new release after changing the YAML header from:
```
output:
    BiocStyle::pdf_document2:
```
to
```
output:
    BiocStyle::pdf_document:
```

to see if that solves anything. Otherwise, I'll just revert to pdf_document2 

All the best,
Kevin

**Comment by kevinrue** [comment]
Tadaaaaaaa! 🎉 🎈 🎂 

@aoles and @mtmorgan 

The build error is now only caused by an independent minor issue: a timeout about checking `bioc-devel mailing list subscription`.
Regarding the Windows build error that I was tracking down, it seems to be be caused by a bad interaction between the `BiocStyle::pdf-document2()` style and Windows.  In contrast, `pdf-document()` or `html_document/2` succesfully build the vignettes on Windows (no problem to report on Linux or OSX).

However, that only means that I know how to avoid the problem, not how to solve it.

All the best,
Kevin

**Comment by mtmorgan** [comment]
Hi @kevinrue the package looks great and I'll mark it as accepted. It will be added to the nightly build system and svn, with additional information provided by email in the next several days. It will be included in the forthcoming Bioconductor release, even if accepted slightly after the deadline. Thanks for your work!

**Comment by kevinrue** [comment]
Thanks @mtmorgan amd @lawremi 

All the best,
Kevin
